gapic-generator/lib/gapic/schema/request_param_parser.rb

```diff
@@ -0,0 +1,109 @@
+# frozen_string_literal: true
+
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "gapic/schema/request_parameter"
+require "gapic/schema/parameter_schema"
+
+module Gapic
+  module Schema
+    # Contains logic for parsing protoc request parameters
+    # from the plugin_opt command line
+    module RequestParamParser
+      class << self
+        # Unescapes a symbol from the string
+        # e.g. "a\.b", '.' => "a.b"
+        # @param string [String] String to unescape
+        # @return [String] Unescaped string
+        def unescape string
+          string ? string.gsub(/\\./) { |escaped| escaped[1] } : string
+        end
+
+        # Splits a string by an unescaped symbol
+        # e.g. "a\.b.c.d\.e", '.'  => ["a\.b","c", "d\.e"]
+        # @param string [String] String to split
+        # @param symbol [String] Symbol to split by
+        # @param max_splits [Integer] Maximum amount of splits to perform; -1 for no limit
+        # @return [Array<String>] List of split string parts
+        def split_by_unescaped string, symbol, max_splits = -1
+          splits = 0
+          string.scan(/\\.|#{symbol}|[^#{symbol}\\]+/).each_with_object([String.new]) do |tok, arr|
```

Comment:
For safety, call

```ruby
escaped_symbol = Regexp.escape symbol
```

and interpolate that into the regex instead. In case we use `.` or other special regex character as the symbol in the future.

Reply:
fixed. nicely spotted, stuff like '^' was doing a number on the parsing. (I hope we won't have to use ^ in the param syntax but who knows)